setup.py

```diff
@@ -1,8 +1,96 @@
 #!/usr/bin/env python
 # -*- coding: utf-8 -*-
 
-from setuptools import setup, find_packages
+import os
+import sys
+from setuptools import setup, find_packages, Extension, Command
+from setuptools.command.build_ext import build_ext
 
+
+# ---------------------------------------------------------------------------- #
+#
+def join_path_list(path_list):
+    path = path_list[0]
+    path_list.remove(path)
+    for _path in path_list:
+        path = os.path.join(path, _path)
+    return path
+
+
+# ---------------------------------------------------------------------------- #
+#
+def TomoPyLibraryFile():
+    """
+    Determines the library output file
+    """
+
+    libname = None
+    if sys.platform.lower().startswith('win'):
+        libname = 'libtomopy.dll'
+    elif sys.platform == 'darwin':
+        libname = 'libtomopy.dylib'
+    else:
+        libname = 'libtomopy.so'
+
+    path = join_path_list([os.getcwd(), "tomopy", "sharedlibs", libname])
+    if os.path.exists(path):
+        return path
+    else:
+        path = join_path_list([os.getcwd(), "tomopy", "sharedlibs", "libtomopy"])
+        for ext in ["so", "dylib", "dll"]:
+            tmp = "{}.{}".format(path, ext)
+            if os.path.exists(tmp):
+                return tmp
+
+    # return best guess
+    return join_path_list([os.getcwd(), "tomopy", "sharedlibs", libname])
+
```

Comment:
This function is very confusing to me. I believe this function is supposed to return the location of the shared library named `libtomopy.*` whether or not it exists. Why does it check for existence if its just going to return a non-existing file anyway?

Reply:
C/C++ habit of always returning the same type -- I return a string because a string is expected. The try-catch block around `shutil.copy2(src, dst)` is missing a re-raise after it printed diagnostic info on where it was looking for the files and folder it was trying to copy to

Reply:
Also, this method covers the three primary errors that will occur in one check: 

  1) file doesn't exist because build failed
  2) insufficient permissions issue causing the copy to fail
  3) the target directory doesn't exist. 

This style enables all of those to be relegated to a single try-catch block without the need for a `if src is not None:` along with the check that the copy succeeded

Reply:
I think a more pythonic way would have been to return `None` if the shared library doesn't exist because trying to copy `None` also raises and exception at `shutil.copy2` that you can catch.

Reply:
Then we end up with:
```python
        try:
            src_base = os.path.basename(src)
            dst = os.path.join(extdir, "tomopy", "sharedlibs", src_base)
            shutil.copy2(src, dst)
        except Exception as e:
            print("Exception occurred copying '{}' to '{}'...".format(src, dst))
            raise e
```
now neither `src` nor `dst` will possibly be set so the error message is much less informative:
```
"Exception occurred copying 'None' to 'None'..."
```

Reply:
And since a failure in `src_base` occurs before `dst` is assigned or even declared, I believe we would then raise an exception within an exception

Reply:
I care less about making things "Pythonic" than I do about making sure that the code is safe and when it does fail, provides actual information about what went wrong instead of just a backtrace to where the error occurred.

Reply:
Oh! I see. You are relying on `shutil.copy2` to return three different error messages for each of the three errors you expect to occur. That's clever.